EIPS/eip-9999.md

```diff
@@ -0,0 +1,150 @@
+
+#### Transaction Processing
+
+```python
+def process_transaction(env: vm.Environment, tx: Transaction) -> Tuple[Uint, Tuple[Log, ...], Optional[Exception]]:
+    preaccessed_addresses = set()
+    preaccessed_storage_keys = set()
+    
+    # Add block-level pre-accessed slots
+    preaccessed_addresses.add(env.block_level_accessed_addresses)
+    preaccessed_storage_keys.add(env.block_level_accessed_storage_keys)
+    
+    # Handle access lists from transaction
+    ...
+```
+
+This adds the block-level-accessed addresses and storage keys to the preaccessed addresses and storage keys.
+As a result, from the perspective of a transaction, block-level accessed addresses and storage keys are treated the same as precompiles or the coinbase address.
+
+```python
+def process_message_call(message: Message, env: Environment) -> MessageCallOutput:
+    return MessageCallOutput(
+        # ... other fields ...
+        accessed_addresses=evm.accessed_addresses,
+        accessed_storage_keys=evm.accessed_storage_keys
+    )
+```
+
+The message call processing tracks accessed addresses and storage keys during execution, which are then propagated back up to the transaction level and ultimately to the block level.
+
+## Rationale
+
+The proposal builds on several key observations:
+
+1. **Caching Efficiency**: Modern Ethereum clients already implement sophisticated caching mechanisms at the block level. Extending address and storage key warming to match this caching behavior better aligns gas costs with actual computational costs.
+
+2. **Backward Compatibility**: The worst-case scenario for any transaction remains unchanged - it will never pay more than the current cold access cost for its first access to a slot.
+
+3. **First Access Warms For All**: Under the proposed mechanism, the first transaction in a block that accesses and warms multiple addresses or storage slots bears the entire cost of warming, even if subsequent transactions benefit from those already-warmed slots without incurring additional costs. This approach is deemed acceptable because early execution within a block—where warming typically occurs—is generally occupied for transactions placed by professional builders aiming for top-of-block execution. Since the resulting cost discrepancy is relatively minor, a more complex cost-sharing model is avoided in favor of maintaining simplicity.
+An alternative to the one-warms-for-all mechanism involves distributing the warming costs across all accesses within a block. In this approach, each transaction must at least be able to pay the cold access cost. Once all transactions in the block are executed, the total warming costs are evenly distributed, and transaction originators are refunded accordingly.
```

Comment:
From an economic perspective this is fair, but:

- This is currently about `(2100 - 100) = 2000` cost for storage and `(2600 - 100) = 2500` for accounts, thus we would have to distribute these 2k/2.5k gas costs over these txs, and those are only per slot/account once
- This seems super complex
- How to deal with accounting errors, for instance 3 accounts pay the 2k cost for a storage? 
- This would likely mean that on the gas accounting when executing, each tx will have to pay the cold gas upfront
- This will thus resort back to the original 2929 pricing where on the tx-level everything is cold, and will thus not lead to measured empirical efficiency of 5%

Reply:
I'm not a fan of the refund-approch neither. To me it feels like a clear pareto improvement even without doing refunds.